### PR TITLE
call __finalize__ in more methods

### PR DESCRIPTION
**Pull request by arw2019**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

The goal here is to make progress on some of the more uncontroversial/easier content of #28283. I'll split off into another PR if this gets too big

### CODE REVIEW

**Comment by jreback** [comment]
cc @TomAugspurger 

lgtm.

File: pandas/core/frame.py [hunk 1]
```diff
@@ -7083,9 +7084,11 @@ def stack(self, level=-1, dropna=True):
         from pandas.core.reshape.reshape import stack, stack_multiple
 
         if isinstance(level, (tuple, list)):
-            return stack_multiple(self, level, dropna=dropna)
+            result = stack_multiple(self, level, dropna=dropna)
         else:
-            return stack(self, level, dropna=dropna)
+            result = stack(self, level, dropna=dropna)
+
+        return result.__finalize__(self, method="stack")
```
Comment (TomAugspurger) [inline pandas/core/frame.py:7091]: I think this falls into the `name` issue I spotted in https://github.com/pandas-dev/pandas/pull/37037. Essentially, if `df` has a column named `name` we try to set `result.name = self.name`, i.e. the series `name`.

```python
In [10]: df = pd.DataFrame({"A": [1, 2], "name": [3, 4]})

In [11]: df.stack().__finalize__(df)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-899f15844169> in <module>
----> 1 df.stack().__finalize__(df)

~/Envs/dask-dev/lib/python3.8/site-packages/pandas/core/generic.py in __finalize__(self, other, method, **kwargs)
   5157         if isinstance(other, NDFrame):
   5158             for name in self._metadata:
-> 5159                 object.__setattr__(self, name, getattr(other, name, None))
   5160         return self
   5161

~/Envs/dask-dev/lib/python3.8/site-packages/pandas/core/series.py in name(self, value)
    464     def name(self, value):
    465         if value is not None and not is_hashable(value):
--> 466             raise TypeError("Series.name must be a hashable type")
    467         object.__setattr__(self, "_name", value)
    468

TypeError: Series.name must be a hashable type
```

I'll think the solution is to only iterate over the intersection of `self._metadata` and `other._metadata`. I'll put up a PR.

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/pull/37205 handles the dataframe.name issue.

I don't know if we want to add a test case for all of those, but one or two might not hurt?

**Comment by arw2019** [comment]
> #37205 handles the dataframe.name issue.
> 
> I don't know if we want to add a test case for all of those, but one or two might not hurt?

Great! 

I can add the tests here and then probably we would want to merge #37205 before this goes in

**Comment by jreback** [comment]
can you merge master and ping on green as merged @TomAugspurger patch

**Comment by arw2019** [comment]
Green

**Comment by jreback** [comment]
thanks @arw2019

**Pull request merged**